File: src/main/java/com/gusrylmubarok/reddit/backend/service/impl/AuthServiceImpl.java
package com.gusrylmubarok.reddit.backend.service.impl;

import com.gusrylmubarok.reddit.backend.dto.request.LoginRequest;
import com.gusrylmubarok.reddit.backend.dto.request.RefreshTokenRequest;
import com.gusrylmubarok.reddit.backend.dto.request.RegisterRequest;
import com.gusrylmubarok.reddit.backend.dto.response.AuthenticationResponse;
import com.gusrylmubarok.reddit.backend.exceptions.ActivationException;
import com.gusrylmubarok.reddit.backend.model.AccountVerificationToken;
import com.gusrylmubarok.reddit.backend.model.NotificationEmail;
import com.gusrylmubarok.reddit.backend.model.User;
import com.gusrylmubarok.reddit.backend.repository.AccountVerificationTokenRepository;
import com.gusrylmubarok.reddit.backend.repository.UserRepository;
import com.gusrylmubarok.reddit.backend.security.JwtProvider;
import com.gusrylmubarok.reddit.backend.service.AuthService;
import com.gusrylmubarok.reddit.backend.service.MailService;
import com.gusrylmubarok.reddit.backend.service.RefreshTokenService;
import lombok.AllArgsConstructor;
import org.springframework.security.authentication.AnonymousAuthenticationToken;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.security.oauth2.jwt.Jwt;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.util.Optional;
import java.util.UUID;

@Service
@AllArgsConstructor
@Transactional
public class AuthServiceImpl implements AuthService {
    private final PasswordEncoder passwordEncoder;
    private final UserRepository userRepository;
    private final AccountVerificationTokenRepository accountVerificationToken;
    private final MailService mailService;
    private final AuthenticationManager authenticationManager;
    private final JwtProvider jwtProvider;
    private final RefreshTokenService refreshTokenService;

    @Override
    public String signup(RegisterRequest registerRequest) {
        User userExists = userRepository.findByEmail(registerRequest.getEmail());
        if (userExists == null) {
            User user = new User();
            user.setUsername(registerRequest.getUsername());
            user.setEmail(registerRequest.getEmail());
            user.setPassword(passwordEncoder.encode(registerRequest.getPassword()));
            user.setCreationDate(Instant.now());
            user.setAccountStatus(false);

            userRepository.save(user);

            String token = generateVerificationToken(user);
            mailService.sendMail(new NotificationEmail("Please Activate your Account",
                    user.getEmail(), "Thank you for signing up to Spring Reddit, " +
                    "please click on the below url to activate your account : " +
                    "http://localhost:8080/api/v1/auth/accountVerification/" + token));
            return "Activation has been send";
        }else {
            if(isActiveUser(userExists.getEmail()) == true) {
                AccountVerificationToken tokenUser = accountVerificationToken.findByUser(userExists);
                String token = tokenUser.getToken();
                mailService.sendMail(new NotificationEmail("Please Activate your Account",
                        userExists.getEmail(), "Thank you for signing up to Spring Reddit, " +
                        "please click on the below url to activate your account : " +
                        "http://localhost:8080/api/v1/auth/accountVerification/" + token));
                return "Activation has been send again";
            }else {
                throw new ActivationException("Email has already used");
            }
        }
    }

    @Override
    public AuthenticationResponse login(LoginRequest loginRequest) {
        Authentication authenticate = authenticationManager.authenticate(new UsernamePasswordAuthenticationToken(loginRequest.getUsername(),
                loginRequest.getPassword()));
        SecurityContextHolder.getContext().setAuthentication(authenticate);
        String token = jwtProvider.generateToken(authenticate);
        return AuthenticationResponse.builder()
                .authenticationToken(token)
                .refreshToken(refreshTokenService.generateRefreshToken().getToken())
                .expiresAt(Instant.now().plusMillis(jwtProvider.getJwtExpirationInMillis()))
                .username(loginRequest.getUsername())
                .build();
    }

    @Override
    @Transactional(readOnly = true)
    public User getCurrentUser() {
        Jwt principal = (Jwt) SecurityContextHolder.
                getContext().getAuthentication().getPrincipal();
        return userRepository.findByUsername(principal.getSubject())
                .orElseThrow(() -> new UsernameNotFoundException("User name not found - " + principal.getSubject()));
    }

    @Override
    public void enableAccount(AccountVerificationToken token) {
        String username = token.getUser().getUsername();
        User user = userRepository.findByUsername(username)
                .orElseThrow(() -> new ActivationException("User not found with username: " + username));
        user.setAccountStatus(true);
        userRepository.save(user);
    }

    @Override
    public String generateVerificationToken(User user) {
        String token = UUID.randomUUID().toString();
        AccountVerificationToken accountVerificationToken = new AccountVerificationToken();
        accountVerificationToken.setToken(token);
        accountVerificationToken.setUser(user);

        this.accountVerificationToken.save(accountVerificationToken);
        return token;
    }

    @Override
    public void verifyAccount(String token) {
        Optional<AccountVerificationToken> verificationToken = accountVerificationToken.findByToken(token);
        verificationToken.orElseThrow(() -> new ActivationException("Invalid Activation Token"));
        enableAccount(verificationToken.get());
    }

    @Override
    public AuthenticationResponse refreshToken(RefreshTokenRequest refreshTokenRequest) {
        refreshTokenService.validateRefreshToken(refreshTokenRequest.getRefreshToken());
        String token = jwtProvider.generateTokenWithUserName(refreshTokenRequest.getUsername());
        return AuthenticationResponse.builder()
                .authenticationToken(token)
                .refreshToken(refreshTokenRequest.getRefreshToken())
                .expiresAt(Instant.now().plusMillis(jwtProvider.getJwtExpirationInMillis()))
                .username(refreshTokenRequest.getUsername())
                .build();
    }

    @Override
    public boolean isLoggedIn() {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        return !(authentication instanceof AnonymousAuthenticationToken) && authentication.isAuthenticated();
    }

    private boolean isActiveUser(String email) {
        User user = userRepository.findByEmail(email);
        return user.isAccountStatus() == true ? true : false;
    }
}

File: src/main/java/com/reddot/service/impl/AuthServiceImpl.kt
package com.reddot.service.impl

import com.reddot.common.Constants
import com.reddot.entity.TokenEmail
import com.reddot.entity.User
import com.reddot.exception.BadRequestException
import com.reddot.mail.MailContentService
import com.reddot.mail.MailServiceImpl
import com.reddot.model.*
import com.reddot.repository.TokenEmailRepository
import com.reddot.repository.UserRepository
import com.reddot.security.UserDetailsImpl
import com.reddot.security.jwt.JwtProvider
import com.reddot.service.AuthService
import com.reddot.service.RefreshTokenService
import com.reddot.validator.ValidationUtil
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.Instant
import java.util.*

@Service
class AuthServiceImpl(
    private val userRepository: UserRepository,
    private val tokenEmailRepository: TokenEmailRepository,
    private val mailContentService: MailContentService,
    private val mailServiceImpl: MailServiceImpl,
    private val passwordEncoder: PasswordEncoder,
    private val validationUtil: ValidationUtil,
    private val authenticationManager: AuthenticationManager,
    private val jwtProvider: JwtProvider,
    private val refreshTokenService: RefreshTokenService
) : AuthService {

    @Transactional
    override fun signup(param: SignupRequest): MessageResponse {
        validationUtil.validate(param)
        if (!availableEmail(param.email)) {
            throw BadRequestException("email already exists")
        }
        if (!availableUsername(param.username)) {
            throw BadRequestException("username already exists")
        }

        val user = User(
            id = null,
            name = param.name,
            username = param.username,
            email = param.email,
            password = encodePassword(param.password),
            enabled = false,
            role = "ROLE_USER",
            createdAt = Date(),
            updatedAt = null
        )
        userRepository.save(user)
        // token
        val token = generateTokenEmail(user)
        val message = mailContentService
            .build(
                "Thank you for registration to Reddot, " +
                        "please click on the below url to activate your account : " +
                        Constants.ACTIVATION_EMAIL + "/" + token
            )
        mailServiceImpl.sendMail(EmailNotify("Please Active Your Account", param.email, message))

        return MessageResponse("Please Activation on ${param.email}")
    }

    @Transactional
    override fun activeAccount(request: String): MessageResponse {
        val token = tokenEmailRepository.findByToken(request) ?: throw BadRequestException("token is expired")
        val user = userRepository.findByUsername(token.user.username) ?: throw BadRequestException("username is not register")
        user.enabled = true
        user.updatedAt = Date()
        userRepository.save(user)

        return MessageResponse("Account has been active ${user.username}")
    }

    @Transactional
    override fun signin(param: SigninRequest): SigninResponse {
        validationUtil.validate(param)

        val auth = authenticationManager.authenticate(UsernamePasswordAuthenticationToken(param.username, param.password))
        if (!auth.isAuthenticated) {
            throw BadRequestException("Wrong username or password")
        }
        SecurityContextHolder.getContext().authentication = auth

        val principal = auth.principal as UserDetailsImpl
        val token = jwtProvider.generateToken(auth)
        val tokenExpired = jwtProvider.getTokenExpirationInMillis()
        val refreshToken = refreshTokenService.generateRefreshToken()

        return SigninResponse(
            username = principal.username,
            token = token,
            expiresAt = Instant.now().plusMillis(tokenExpired),
            refreshToken = refreshToken.token
        )
    }

    @Transactional
    override fun refreshToken(param: RefreshTokenRequest): SigninResponse {
        validationUtil.validate(param)
        refreshTokenService.validateRefreshToken(param.refreshToken)

        val newToken = jwtProvider.generateRefreshToken(param.username)
        val tokenExpired = jwtProvider.getTokenExpirationInMillis()

        return SigninResponse(
            username = param.username,
            token = newToken,
            expiresAt = Instant.now().plusMillis(tokenExpired),
            refreshToken = param.refreshToken
        )
    }

    private fun encodePassword(password: String): String {
        return passwordEncoder.encode(password)
    }

    private fun generateTokenEmail(user: User): String {
        val token = UUID.randomUUID().toString()
        val tokenEmail = TokenEmail(
            id = null,
            token = token,
            user = user,
            expiryDate = Instant.now()
        )
        tokenEmailRepository.save(tokenEmail)
        return token
    }

    private fun availableEmail(email: String) : Boolean {
        userRepository.findByEmail(email) ?: return true
        return false
    }

    private fun availableUsername(username: String) : Boolean {
        userRepository.findByUsername(username) ?: return true
        return false
    }
}